tensorboard/data/provider.py

```diff
@@ -0,0 +1,223 @@
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Experimental framework for generic TensorBoard data providers."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import abc
+import collections
+
+import six
+
+
+@six.add_metaclass(abc.ABCMeta)
+class DataProvider(object):
+  @abc.abstractmethod
+  def list_scalars(
+      self,
+      experiment_id,
+      owner_plugin,
+      run_tag_filter=None,
+  ):
+    """List metadata about scalar time series.
+
+    Args:
+      experiment_id: ID of enclosing experiment.
+      owner_plugin: String name of the TensorBoard plugin that created
+        the data to be queried. Required.
+      run_tag_filter: Optional `RunTagFilter` value. If omitted, all
+        runs and tags will be included.
+
+    The result will only contain keys for run-tag combinations that
+    actually exist, which may not include all entries in the
+    `run_tag_filter`.
+
+    Returns:
+      A nested map `d` such that `d[run][tag]` is a `ScalarMetadata`
+      value.
+    """
+    pass
+
+  @abc.abstractmethod
+  def read_scalars(
+      self,
+      experiment_id,
+      owner_plugin,
+      downsample_to=None,
+      run_tag_filter=None,
+      step_filter=None,
+  ):
+    """Read values from scalar time series.
+
+    Args:
+      experiment_id: ID of enclosing experiment.
+      owner_plugin: String name of the TensorBoard plugin that created
+        the data to be queried. Required.
+      downsample_to: Integer number of steps to which to downsample the
+        results (e.g., `1000`). Required.
+      run_tag_filter: Optional `RunTagFilter` value. If provided, a time
+        series will only be included in the result if its run and tag
+        both pass this filter. If `None`, all time series will be
+        included.
+      step_filter: Optional `StepFilter` value. If `None`, the entire
+        range of steps may be included.
+
+    The result will only contain keys for run-tag combinations that
+    actually exist, which may not include all entries in the
+    `run_tag_filter`.
+
+    Returns:
+      A nested map `d` such that `d[run][tag]` is a list of
+      `ScalarDatum` values sorted by step.
+    """
+    pass
+
+  def list_tensors(self):
+    """Not yet specified."""
+    pass
+
+  def read_tensors(self):
+    """Not yet specified."""
+    pass
+
+  def list_blob_sequences(self):
+    """Not yet specified."""
+    pass
+
+  def read_blob_sequences(self):
+    """Not yet specified."""
+    pass
+
+
+_ScalarMetadata = collections.namedtuple(
```

Comment:
Thinking about this a little more, maybe we should avoid namedtuple here.  The main reason is that we don't really need the tuple behavior, and having it will encourage uses that will make it hard to add new fields later, specifically if people do:

`max_step, wall_time, metadata = scalar_metadata`

This is expounded on in a bit more length at https://news.ycombinator.com/item?id=15132670.

If we just use `@property` and write out the class, we can still make it read-only so all we're really losing is `__slots__` which we could add by hand (even if it's a bit uglier then).  We don't get a nice repr and equality behavior, but we're already foregoing that for RunTagFilter and StepFilter and that seems fine, right?  And it would be nice if all the objects in the API used a consistent pattern.

Reply:
Sure, makes sense to me. This also avoids the common string formatting
errors (`"My datum: %s" % my_datum` is an error with `namedtuple`s).

All I really want is a simple data type. I’ll go ahead and write out the
whole thing.